packages/data-broker/src/types/models.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2022 Beyond Essential Systems Pty Ltd
+ */
+
+import type {
+  DatabaseModel as BaseDatabaseModel,
+  DatabaseType as BaseDatabaseType,
+  DataSourceModel as BaseDataSourceModel,
+  DataSourceType as BaseDataSourceType,
+  EntityModel as BaseEntityModel,
+  EntityType as BaseEntityType,
+  ModelRegistry,
+} from '@tupaia/database';
+import { Join, Override, Values } from './utils';
+
+type DbValue = string | number | boolean | null | DbValue[] | { [key: string]: DbValue };
+
+export type DbRecord = Record<string, DbValue>;
+
+// Conditions for flat i.e. non array/json fields
+type FlatFieldConditions<T> = {
+  [K in keyof T as T[K] extends Record<string, unknown>
+    ? never
+    : T[K] extends unknown[]
+    ? never
+    : K]: T[K] | T[K][];
+};
+
+type JsonFieldConditions<T> = Values<
+  {
+    [K in keyof T as T[K] extends Record<string, string | number | boolean> ? K : never]: {
+      [K1 in keyof T[K] as Join<K, K1, '->>'>]: T[K][K1] | T[K][K1][];
```

Comment:
Nice one. Looks like we probably want to move this to @tupaia/database or @tupaia/types at some point.

Reply:
Yeap, or even better convert `@tupaia/database` to TS too :wink:

Reply:
Yeah that's fine about not re-using the types in `entity-server`, thanks for checking. They are slightly different. Ideally they'd both sit in `@tupaia/database` as you say 👍 

Just wanna note that if we do consider re-writing `@tupaia/database` in typescript, I'd love if we took the chance to seriously consider switching to a higher level ORM which has inbuilt typescript support. While I've always been impressed by the stability of our custom ORM; it's a lot of code, it quite hard to read through, and it could be more efficient/have more utility functions.

Reply:
Good point @rohan-bes ! I have tried https://www.prisma.io/ in a personal project and I was impressed by its TS support. I cannot do _everything_ (e.g. it doesn't support dynamic types for its CRUD operations - see [this GitHub issue](https://github.com/prisma/prisma/issues/5273)) but it's a great starting point.